src/bst.py

```diff
@@ -0,0 +1,285 @@
+                self.right._insert(val, data)
+            else:
+                self.right = Node(val)
+        ld = self.left.depth if self.left else 0
+        rd = self.right.depth if self.right else 0
+        self.depth = 1 + max(ld, rd)
+
+    def _in_order(self):
+        """
+        This internal method is a generator that will output in order traversal
+        of a binary tree(left child, parent, right child), one value at a time.
+        """
+        try:
+            for node in self.left._in_order():
+                if node is None:
+                    pass
+                else:
+                    yield node
+        except AttributeError:
+            pass
+        yield self.val
+        try:
+            for node in self.right._in_order():
+                if node is None:
+                    pass
+                else:
+                    yield node
+        except AttributeError:
+            pass
+
+    def _pre_order(self):
+        """
+        This internal method is a generator that will output preorder traversal
+        of a binary tree(parent, left child, right child), one value at a time.
+        """
+        yield self.val
+        try:
+            for node in self.left._pre_order():
+                if node is None:
+                    pass
+                else:
+                    yield node
+        except AttributeError:
+            pass
+        try:
+            for node in self.right._pre_order():
+                if node is None:
+                    pass
+                else:
+                    yield node
+        except AttributeError:
+            pass
+
+    def _post_order(self):
+        """
+        This internal method is a generator that will output postorder traversal
+        of a binary tree(left child, right child, parent), one value at a time.
+        """
+        try:
+            for node in self.left._post_order():
+                if node is None:
+                    pass
+                else:
+                    yield node
+        except AttributeError:
+            pass
+        try:
+            for node in self.right._post_order():
+                if node is None:
+                    pass
+                else:
+                    yield node
+        except AttributeError:
+            pass
+        yield self.val
+
+
+class BinarySearchTree(object):
+    """BinarySearchTree implements a Binary Search Tree data structure
+    and associated methods."""
+
+    def __init__(self, root=None):
+        """Create an instance of our Binary Search Tree w/ supplied
+        input, or an emptry tree."""
+        self.length = 0
+        if root:
+            self.root = Node(root)
+            self.length = 1
+        else:
+            self.root = root
+
+    def insert(self, val, data=None):
+        """Inserts a value  into the BST.  If the value is already in
+        the BST it will be ingored."""
+        if not self.contains(val):
```

Comment:
This will work; however you are substantially increasing the time it takes to run `insert`, since you are first traversing the tree to see if the item is contained, then traversing the tree again to insert the item. You can improve on this and do it all in one run through.